modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java

```diff
@@ -409,8 +410,14 @@ public void parse(ParseContext context) throws IOException {
 
         Version indexVersion = context.mapperService().getIndexSettings().getIndexVersionCreated();
         createQueryBuilderField(indexVersion, queryBuilderField, queryBuilder, context);
-        Query query = toQuery(queryShardContext, isMapUnmappedFieldAsText(), queryBuilder);
-        processQuery(query, context);
+        QueryBuilder queryBuilderForProcessing = rewrite(queryBuilder);
```

Comment:
I wonder if we could use `QueryShardContext#isCachable` to determine if the query should be processed at index time or not. We can build the Lucene `Query` all the time and checks if the shard context is cacheable afterward. This way we don't need to visit the query in the `rewrite` function ?

Reply:
Yes, I think that would work. However the downside is that then for the entire query we will not extract terms / ranges. In the current approach only the clause with a date range with now gets ignored, so that we can still benefit from index time terms / range extraction. I think that most of the time these clauses with `now` date ranges, co-exists with other clauses for which we can extract terms/ranges.

This approach does make query processing more complex and the `QueryShardContext#isCachable` is simple. Maybe we should do this and if we encounter feedback that these percolator queries with now date ranges are seen as slow then we can always bring back the approach in this PR?

Reply:
Right I missed the recursion ;). What about Simon's idea: https://github.com/elastic/elasticsearch/pull/35160#discussion_r230829998 ? It is similar but the other clauses would still be extracted so it should be as fast as before.